src/google-classroom-extractor/google_classroom_extractor/api/coursework.py

```diff
@@ -170,18 +176,56 @@ def request_all_coursework_as_df(
     """
 
     coursework_df: DataFrame = request_latest_coursework_as_df(resource, course_ids)
-
-    # append everything from API call
-    coursework_df.to_sql(
-        "Coursework", sync_db, if_exists="append", index=False, chunksize=500
-    )
-    # remove duplicates - leave only the most recent
-    with sync_db.connect() as con:
-        con.execute(
-            "DELETE from Coursework "
-            "WHERE rowid not in (select max(rowid) "
-            "FROM Coursework "
-            "GROUP BY id)"
-        )
+    _sync_without_cleanup(coursework_df, sync_db)
+    cleanup_after_sync(ASSIGNMENTS_RESOURCE_NAME, sync_db)
 
     return coursework_df
+
+
+def _sync_without_cleanup(
+    resource_df: DataFrame, sync_db: sqlalchemy.engine.base.Engine
+):
+    """
+    Take fetched API data and sync with database. Creates tables when necessary,
+    but ok if temporary tables are there to start. Doesn't delete temporary tables when finished.
+
+    Parameters
+    ----------
+    resource_df: DataFrame
+        a courseworks API DataFrame with the current fetched data which
+        will be mutated, adding Hash and CreateDate/LastModifiedDate
+    sync_db: sqlalchemy.engine.base.Engine
+        an Engine instance for creating database connections
+    """
+    sync_to_db_without_cleanup(
+        resource_df=resource_df,
+        resource_name=ASSIGNMENTS_RESOURCE_NAME,
+        table_columns_sql="""
+            courseId TEXT,
+            id TEXT,
+            title TEXT,
+            description TEXT,
+            materials TEXT,
+            state TEXT,
+            alternateLink TEXT,
+            creationTime TEXT,
+            updateTime TEXT,
+            maxPoints TEXT,
+            workType TEXT,
+            submissionModificationMode TEXT,
+            assigneeMode TEXT,
+            creatorUserId TEXT,
+            "dueDate.year" TEXT,
+            "dueDate.month" TEXT,
+            "dueDate.day" TEXT,
+            "dueTime.hours" TEXT,
+            "dueTime.minutes" TEXT,
+            "assignment.studentWorkFolder.id" TEXT,
+            "assignment.studentWorkFolder.title" TEXT,
+            "assignment.studentWorkFolder.alternateLink" TEXT,
+            topicId TEXT,
+            scheduledTime TEXT,
+            """,
+        primary_keys="id",
```

Comment:
So, "id" probably isn't enough for identity for any of the other tickets.  In this case, it's courseId and id together because id is only unique per course.  See https://developers.google.com/classroom/reference/rest/v1/courses.courseWork
You can cheat by looking at the mapping code where I join id fields with "-" to see which ones are needed for uniqueness.

Reply:
Thank you